File: gte/nodes/turbocompressor/rotor/rotor.go
package rotor

import (
	"fmt"
	"math"

	"github.com/ParkhomenkoDV/gte/gte/metrics"
	"github.com/ParkhomenkoDV/gte/gte/utils"
	su "github.com/ParkhomenkoDV/substance/substance"
	td "github.com/ParkhomenkoDV/thermodynamics/thermodynamics"
	"gonum.org/v1/gonum/diff/fd"
	"gonum.org/v1/gonum/mat"
	"gonum.org/v1/gonum/optimize"
)

type Parameters struct {
	EffEff float64 `doc:"Адиабатический КПД"`
	TiTi   float64 `doc:"Степень повышения полной температуры"`
	PiPi   float64 `doc:"Степень повышения полного давления"`
}

func NewParameters() Parameters {
	return Parameters{}
}

type Rotor struct {
	Name string
	Parameters
}

func (r *Rotor) New() Rotor {
	return Rotor{}
}

// power = m * hcp * (T*_outlet - T*_inlet)
// T*_outlet = T*_inlet * (1 + (pi* ** ((k-1) / k) - 1) / eff*)
// pi* = P*_outlet / P*_inlet
func (r *Rotor) Equations(x []float64, inlet, outlet *su.Substance) []float64 {
	outlet_TT, outlet_PP := x[0], x[1]

	ranges := map[string][2]float64{
		"TT": {inlet.Parameters["TT"], outlet_TT},
		"PP": {inlet.Parameters["PP"], outlet_PP},
		"EO": {inlet.Parameters["EO"], outlet.Parameters["EO"]},
	}

	gc, _ := utils.IntegralAverage(inlet.Functions["gc"], ranges)
	hcp, _ := utils.IntegralAverage(inlet.Functions["hcp"], ranges)
	k := td.AdiabaticIndex(gc, hcp)

	return []float64{
		outlet_TT - inlet.Parameters["TT"]*(1+(math.Pow(r.PiPi, ((k-1)/k)-1))/r.EffEff),
		r.TiTi - outlet_TT/inlet.Parameters["TT"],
		r.PiPi - outlet_PP/inlet.Parameters["PP"],
	}
}

func (r *Rotor) NVars() int {
	return 2
}

func (r *Rotor) IsSolvable(x []float64) bool {
	return false
}

func (r *Rotor) Calculate(inlets ...*su.Substance) ([]*su.Substance, error) {
	if len(inlets) != 1 {
		return []*su.Substance{}, fmt.Errorf("len(inlets)=%v must be = 1", len(inlets))
	}
	inlet := inlets[0]

	outlet := &su.Substance{
		Name:        inlet.Name,
		Composition: inlet.Composition,
		Parameters: su.Parameters{
			"m": inlet.P("m"),
		},
		Functions: inlet.Functions,
	}

	if _, ok := inlet.Parameters["eo"]; ok {
		outlet.Parameters["oxidizer"] = inlet.Parameters["oxidizer"]
		outlet.Parameters["eo"] = inlet.Parameters["eo"]
	}

	loss := func(x []float64) float64 {
		res := r.Equations(x, inlet, outlet)
		return metrics.MSE(res...)
	}

	// Численное вычисление градиента
	grad := func(grad, x []float64) {
		// Используем численное дифференцирование
		fd.Gradient(grad, loss, x, &fd.Settings{
			Formula:    fd.Central, // Центральные разности (наиболее точные)
			Step:       1e-8,       // Шаг дифференцирования
			Concurrent: true,       // Параллельное вычисление
		})
	}

	hess := func(hess *mat.SymDense, x []float64) {
		// Используем численное дифференцирование
		fd.Hessian(hess, loss, x, &fd.Settings{
			Formula:    fd.Central, // Центральные разности (наиболее точные)
			Step:       1e-8,       // Шаг дифференцирования
			Concurrent: true,       // Параллельное вычисление
		})
	}

	problem := optimize.Problem{Func: loss, Grad: grad, Hess: hess}

	k_i := inlet.Parameters["k"]

	if r.EffEff != 0 && r.TiTi != 0 {
		outlet.Parameters["TT"] = inlet.Parameters["TT"] * r.TiTi
		outlet.Parameters["PP"] = inlet.Parameters["PP"] * math.Pow((outlet.Parameters["TT"]/inlet.Parameters["TT"]-1)*r.EffEff+1, k_i/(k_i-1))
	} else if r.EffEff != 0 && r.PiPi != 0 {
		outlet.Parameters["TT"] = inlet.Parameters["TT"] * (1 + (math.Pow(r.PiPi, (k_i-1)/k_i)-1)/r.EffEff)
		outlet.Parameters["PP"] = inlet.Parameters["PP"] * r.PiPi
	} else if r.TiTi != 0 && r.PiPi != 0 {
		outlet.Parameters["TT"] = inlet.Parameters["TT"] * r.TiTi
		outlet.Parameters["PP"] = inlet.Parameters["PP"] * r.PiPi
	}
	initX := []float64{outlet.Parameters["TT"], outlet.Parameters["PP"]}

	settings := &optimize.Settings{
		MajorIterations:   1_000, // Максимальное число итераций
		GradientThreshold: 1e-8,  // Порог для градиента
		Concurrent:        0,     // Автоматический выбор числа потоков
	}

	// Выбираем метод. NelderMead не требует градиентов, прост и надежен.
	// Для более сложных задач используйте &optimize.LBFGS{}
	//method := &optimize.NelderMead{}
	method := &optimize.LBFGS{}

	result, err := optimize.Minimize(problem, initX, settings, method)

	outlet.Parameters["TT"], outlet.Parameters["PP"] = result.X[0], result.X[1]

	return []*su.Substance{outlet}, err
}

func (r *Rotor) Validate() {

}

func (r *Rotor) CheckReal() {

}

File: gte/nodes/turbocompressor/rotor/rotor.py
from typing import Any, Dict, Tuple, Union

from numpy import isnan, nan
from scipy.optimize import root
from substance import Substance
from thermodynamics import adiabatic_index

try:
    from ....checks import check_mass_flow, check_temperature
    from ....config.config import EPSREL
    from ....config.config import parameters as gtep
    from ....errors import TYPE_ERROR
    from ....utils.utils import integral_average
    from ...node import Node
except ImportError:
    import os
    import sys

    sys.path.insert(0, os.getcwd())

    from gte.checks import check_mass_flow, check_temperature
    from gte.config.config import EPSREL
    from gte.config.config import parameters as gtep
    from gte.errors import TYPE_ERROR
    from gte.nodes.node import Node
    from gte.utils.utils import integral_average


class Rotor(Node):
    """Ротор"""

    variables: Tuple[str, str, str] = (gtep.effeff, gtep.titi, gtep.pipi)
    n_vars: int = 2

    __slots__ = ()  # нет новых атрибутов

    def __init__(self, parameters: Dict[str, float], name: str = "Rotor"):
        Node.__init__(self, parameters, name)

    @classmethod
    def _equations(cls, x: Tuple[float], args: Dict[str, Any]) -> Tuple[float, float, float]:
        """
        T*_outlet = T*_inlet * (1 + (pi* ** ((k-1) / k) - 1) / eff*)
        ti* = T*_outlet / T*_inlet
        pi* = P*_outlet / P*_inlet
        """
        outlet_TT, outlet_PP = x[0], x[1]
        effeff, titi, pipi = args.get(gtep.effeff, x[2]), args.get(gtep.titi, x[2]), args.get(gtep.pipi, x[2])

        inlet, outlet = args["inlet"], args["outlet"]

        ranges = {
            gtep.TT: (inlet.parameters[gtep.TT], outlet_TT),
            gtep.PP: (inlet.parameters[gtep.PP], outlet_PP),
            gtep.eo: (inlet.parameters.get(gtep.eo, 0), outlet.parameters.get(gtep.eo, 0)),
        }
        gc, _ = integral_average(inlet.functions[gtep.gc], **ranges)
        hcp, _ = integral_average(inlet.functions[gtep.hcp], **ranges)
        k = adiabatic_index(gc, hcp)

        return (
            outlet_TT - inlet.parameters[gtep.TT] * (1 + (pipi ** ((k - 1) / k) - 1) / effeff),
            titi - outlet_TT / inlet.parameters[gtep.TT],
            pipi - outlet_PP / inlet.parameters[gtep.PP],
        )

    @classmethod
    def predict(cls, parameters: Dict[str, Union[float, int]], inlet: Substance) -> Tuple[Dict[str, float], Substance]:
        """Начальные приближения"""
        Node.validate_substance(inlet)

        if not isinstance(parameters, dict):
            raise TypeError(TYPE_ERROR.format(f"{type(parameters)=}", dict))
        if len(parameters) != cls.n_vars:
            raise ValueError(f"{len(parameters)=} must be {cls.n_vars}")
        for parameter, value in parameters.items():
            if parameter not in cls.variables:
                raise ValueError(f"{parameter=} not in {cls.variables}")
            if not isinstance(value, (float, int)):
                raise TypeError(TYPE_ERROR.format(f"{type(value)=}", float))

        gc_i: float = inlet.functions[gtep.gc](inlet.parameters)
        hcp_i: float = inlet.functions[gtep.hcp](inlet.parameters)
        k_i: float = adiabatic_index(gc_i, hcp_i)

        outlet = Substance(
            inlet.name,
            inlet.composition,
            parameters={
                gtep.m: inlet.parameters[gtep.m],
            },
            functions=inlet.functions,
        )
        if gtep.eo in inlet.parameters:
            outlet.parameters["oxidizer"] = inlet.parameters["oxidizer"]
            outlet.parameters[gtep.eo] = inlet.parameters[gtep.eo]

        if gtep.titi in parameters and gtep.pipi in parameters:
            outlet.parameters[gtep.TT] = inlet.parameters[gtep.TT] * parameters[gtep.titi]
            outlet.parameters[gtep.PP] = inlet.parameters[gtep.PP] * parameters[gtep.pipi]
        elif gtep.effeff in parameters and gtep.titi in parameters:
            outlet.parameters[gtep.TT] = inlet.parameters[gtep.TT] * parameters[gtep.titi]
            outlet.parameters[gtep.PP] = inlet.parameters[gtep.PP] * ((outlet.parameters[gtep.TT] / inlet.parameters[gtep.TT] - 1) * parameters[gtep.effeff] + 1) ** (k_i / (k_i - 1))
        elif gtep.effeff in parameters and gtep.pipi in parameters:
            outlet.parameters[gtep.TT] = inlet.parameters[gtep.TT] * (1 + (parameters[gtep.pipi] ** ((k_i - 1) / k_i) - 1) / parameters[gtep.effeff])
            outlet.parameters[gtep.PP] = inlet.parameters[gtep.PP] * parameters[gtep.pipi]
        else:
            raise ArithmeticError(f"{parameters=}")

        outlet = Node.calculate_substance(outlet)

        return {gtep.effeff: cls.total_efficiency(inlet, outlet), gtep.titi: cls.total_temperature_ratio(inlet, outlet), gtep.pipi: cls.total_pressure_ratio(inlet, outlet)}, outlet

    @classmethod
    def calculate(cls, parameters: Dict[str, Union[float, int]], inlet: Substance) -> Tuple[Dict[str, float], Substance]:
        prediction, outlet_ = cls.predict(parameters, inlet)

        outlet = Substance(
            inlet.name,
            inlet.composition,
            parameters={
                gtep.m: inlet.parameters[gtep.m],
            },
            functions=inlet.functions,
        )

        if gtep.eo in inlet.parameters:
            outlet.parameters["oxidizer"] = inlet.parameters["oxidizer"]
            outlet.parameters[gtep.eo] = inlet.parameters[gtep.eo]

        args: Dict[str, Any] = {"inlet": inlet, "outlet": outlet, **parameters}  # НУ
        x0 = [outlet_.parameters[gtep.TT], outlet_.parameters[gtep.PP]] + [prediction[v] for v in cls.variables if v not in parameters]

        result = root(cls._equations, x0, args, method="hybr")

        outlet.parameters[gtep.TT], outlet.parameters[gtep.PP] = float(result.x[0]), float(result.x[1])
        outlet = Node.calculate_substance(outlet)

        effeff = parameters.get(gtep.effeff, cls.total_efficiency(inlet, outlet))
        titi = parameters.get(gtep.titi, cls.total_temperature_ratio(inlet, outlet))
        pipi = parameters.get(gtep.pipi, cls.total_pressure_ratio(inlet, outlet))

        return {gtep.effeff: effeff, gtep.titi: titi, gtep.pipi: pipi, gtep.power: cls.power(inlet, outlet)}, outlet

    @classmethod
    def validate(cls, inlet: Substance, outlet: Substance, epsrel: float = EPSREL) -> Dict[int, float]:
        effeff = cls.total_efficiency(inlet, outlet)
        titi = cls.total_temperature_ratio(inlet, outlet)
        pipi = cls.total_pressure_ratio(inlet, outlet)

        x0 = (outlet.parameters[gtep.TT], outlet.parameters[gtep.PP], effeff, titi, pipi)
        args = {"inlet": inlet, "outlet": outlet, gtep.effeff: effeff, gtep.titi: titi, gtep.pipi: pipi}

        result: Dict[int, float] = {}
        for i, null in enumerate(cls._equations(x0, args)):
            if isnan(null) or abs(null) > epsrel:
                result[i] = null

        return result

    @classmethod
    def check_real(cls, inlet: Substance, outlet: Substance) -> str:
        if not check_mass_flow(inlet.parameters[gtep.m]):
            return f"inlet {gtep.m} {inlet.parameters[gtep.m]}"
        if not check_mass_flow(outlet.parameters[gtep.m]):
            return f"outlet {gtep.m} {outlet.parameters[gtep.m]}"

        if not check_temperature(inlet.parameters[gtep.TT]):
            return f"inlet {gtep.TT} {inlet.parameters[gtep.TT]}"
        if not check_temperature(outlet.parameters[gtep.TT]):
            return f"outlet {gtep.TT} {outlet.parameters[gtep.TT]}"

        return ""

    @classmethod
    def total_efficiency(cls, inlet: Substance, outlet: Substance) -> float:
        """Адиабатический КПД"""
        if not isinstance(inlet, Substance):
            raise TypeError(TYPE_ERROR.format(f"{type(inlet)=}", Substance))
        if not isinstance(outlet, Substance):
            raise TypeError(TYPE_ERROR.format(f"{type(outlet)=}", Substance))

        titi = outlet.parameters[gtep.TT] / inlet.parameters[gtep.TT]
        pipi = cls.total_pressure_ratio(inlet, outlet)

        ranges = {
            gtep.TT: (inlet.parameters[gtep.TT], outlet.parameters[gtep.TT]),
            gtep.PP: (inlet.parameters[gtep.PP], outlet.parameters[gtep.PP]),
            gtep.eo: (inlet.parameters.get(gtep.eo, 0), outlet.parameters.get(gtep.eo, 0)),
        }
        gc, _ = integral_average(inlet.functions[gtep.gc], **ranges)
        hcp, _ = integral_average(inlet.functions[gtep.hcp], **ranges)
        k = adiabatic_index(gc, hcp)

        return (pipi ** ((k - 1) / k) - 1) / (titi - 1)

    @classmethod
    def total_temperature_ratio(cls, inlet: Substance, outlet: Substance) -> float:
        """Степень повышения полной температуры"""
        if not isinstance(inlet, Substance):
            raise TypeError(TYPE_ERROR.format(f"{type(inlet)=}", Substance))
        if not isinstance(outlet, Substance):
            raise TypeError(TYPE_ERROR.format(f"{type(outlet)=}", Substance))

        return outlet.parameters.get(gtep.TT, nan) / inlet.parameters.get(gtep.TT, nan)

    @classmethod
    def total_pressure_ratio(cls, inlet: Substance, outlet: Substance) -> float:
        """Степень повышения полного давления"""
        if not isinstance(inlet, Substance):
            raise TypeError(TYPE_ERROR.format(f"{type(inlet)=}", Substance))
        if not isinstance(outlet, Substance):
            raise TypeError(TYPE_ERROR.format(f"{type(outlet)=}", Substance))

        return outlet.parameters.get(gtep.PP, nan) / inlet.parameters.get(gtep.PP, nan)

    @classmethod
    def power(cls, inlet: Substance, outlet: Substance) -> float:
        """Мощность"""
        if not isinstance(inlet, Substance):
            raise TypeError(TYPE_ERROR.format(f"{type(inlet)=}", Substance))
        if not isinstance(outlet, Substance):
            raise TypeError(TYPE_ERROR.format(f"{type(outlet)=}", Substance))

        ranges = {
            gtep.TT: (inlet.parameters[gtep.TT], outlet.parameters[gtep.TT]),
            gtep.PP: (inlet.parameters[gtep.PP], outlet.parameters[gtep.PP]),
            gtep.eo: (inlet.parameters.get(gtep.eo, 0), outlet.parameters.get(gtep.eo, 0)),
        }
        hcp, _ = integral_average(inlet.functions[gtep.hcp], **ranges)

        return inlet.parameters[gtep.m] * hcp * (outlet.parameters[gtep.TT] - inlet.parameters[gtep.TT])


if __name__ == "__main__":
    from gte.fixtures import air, exhaust

    test_cases = (
        # compressor
        {"parameters": {gtep.pipi: 6, gtep.effeff: 0.85}, "inlet": air},
        {"parameters": {gtep.pipi: 6, gtep.titi: 1.775}, "inlet": air},
        {"parameters": {gtep.effeff: 0.85, gtep.titi: 1.775}, "inlet": air},
        # turbine
        {"parameters": {gtep.effeff: 1 / 0.9, gtep.titi: 1 / 1.775}, "inlet": exhaust},
        {"parameters": {gtep.pipi: 1 / 4.0, gtep.titi: 1 / 1.775}, "inlet": exhaust},
        {"parameters": {gtep.effeff: 1 / 0.9, gtep.pipi: 1 / 4.0}, "inlet": exhaust},
    )
    for test_case in test_cases:
        r = Rotor(test_case["parameters"], "test")
        print(f"{r.is_solvable=}")

        vars, outlet = r.calculate(r.parameters, test_case["inlet"])

        for k, v in outlet.parameters.items():
            print(f"{k:25}: {v:.4f}")
        print(vars)

        print(f"{Rotor.validate(test_case["inlet"], outlet) = }")
        print(f"{Rotor.check_real(test_case["inlet"], outlet) = }")
        print()
